Maximum cardinality matching in bipartite graphs.

jfinkels opened the pull request:
Adds the Hopcroft--Karp algorithm for finding a maximum cardinality
matching in bipartite graphs.

I wasn't sure whether to put this in `networkx.algorithms.matching` or in `networkx.algorithms.bipartite`; I chose the latter.

jfinkels commented:
The previous suggested changes failed on Python2.7. I have updated the pull request to resolve that issue.

jfinkels commented:
The Travis CI errors seem to be from an unrelated unit test, `test_timing`. If you examine the build log and search for the string `Hopcroft--Karp`, you'll see that the test passes.

hagberg commented:
Yes, you are right.  Those tests are not related to your code but instead in the timing tests.  @dschult can you look at this?  Do we need to run these tests regularly or are they primarily for evaluation performance  when suggesting changes?

hagberg commented:
@jfinkels I took at look at the code and I see it follows the wikipedia article as you mention.  I was looking at the Eppstein implementation http://www.ics.uci.edu/~eppstein/PADS/BipartiteMatching.py which is quite different.  Do you have comments on the differences?

jfinkels commented:
I'm not familiar enough with the algorithm to say, but I trust that Professor Eppstein's implementation is as good if not better than mine (or Wikipedia's). It seems to be the same general idea done in a slightly different way, but it's hard for me to say for certain without spending a lot of time trying to interpret the code in both implementations.

I can add his implementation as well, choose a reasonable default implementation to present to the user, and then allow the user to explicitly choose one or the other in case one is more efficient.

jfinkels commented:
I've just updated this pull request again; this time I have two implementations of the Hopcroft--Karp maximum matching algorithm for bipartite graphs: one from Wikipedia and one from David Eppstein's PADS. I've created a module-level attribute `networkx.bipartite.maximum_matching` which is simply an alias for one of the algorithms; users who don't care about implementation should use this one. Finally, I have added an additional function, which computes the minimum vertex cover corresponding to the maximum matching computed by (either of) the maximum matching functions.

ysitu commented on networkx/algorithms/bipartite/matching.py:
```diff
@@ -0,0 +1,422 @@
+            else:
+                distances[v] = INFINITY
+        distances[None] = INFINITY
+        while len(queue) > 0:
```
`while queue:`

ysitu commented on networkx/algorithms/bipartite/matching.py:
```diff
@@ -0,0 +1,422 @@
+                distances[v] = INFINITY
+        distances[None] = INFINITY
+        while len(queue) > 0:
+            v = queue.pop(0)
```
Use `collections.deque` and `popleft` to avoid quadratic behavior.

jfinkels commented:
I have made the changes suggested by @ysitu; specifically using an instance of `collections.deque` instead of a plain list to use as a queue.

ysitu commented on networkx/algorithms/bipartite/matching.py:
```diff
@@ -0,0 +1,423 @@
+        def recurse(v):
+            if v in preds:
+                L = preds[v]
+                del preds[v]
```
`L = preds.pop(v)`

ysitu commented on networkx/algorithms/bipartite/matching.py:
```diff
@@ -0,0 +1,423 @@
+                for u in L:
+                    if u in pred:
+                        pu = pred[u]
+                        del pred[u]
```
`pu = pred.pop(u)`

jfinkels commented:
@ysitu Fixed again.
